ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -101,38 +103,47 @@ initialChainSelection ::
      )
   => ImmutableDB m blk
   -> VolatileDB m blk
-  -> LgrDB m blk
+  -> LedgerDB' m blk
+  -> ResourceRegistry m
   -> Tracer m (TraceInitChainSelEvent blk)
   -> TopLevelConfig blk
   -> StrictTVar m (WithFingerprint (InvalidBlocks blk))
   -> StrictTVar m (FutureBlocks m blk)
   -> CheckInFuture m blk
   -> LoE ()
-  -> m (ChainAndLedger blk)
-initialChainSelection immutableDB volatileDB lgrDB tracer cfg varInvalid
+  -> m (ChainAndLedger m blk)
+initialChainSelection immutableDB volatileDB lgrDB rr tracer cfg varInvalid
                       varFutureBlocks futureCheck loE = do
     -- We follow the steps from section "## Initialization" in ChainDB.md
 
-    (i :: Anchor blk, succsOf, ledger) <- atomically $ do
+    (i :: Anchor blk, succsOf) <- atomically $ do
       invalid <- forgetFingerprint <$> readTVar varInvalid
-      (,,)
+      (,)
         <$> ImmutableDB.getTipAnchor immutableDB
         <*> (ignoreInvalidSuc volatileDB invalid <$>
               VolatileDB.filterByPredecessor volatileDB)
-        <*> LgrDB.getCurrent lgrDB
+
+    -- This is safe: the LedgerDB tip doesn't change in between the previous
+    -- atomically block and this call to 'withTipForker'.
+    --
+    -- We don't use 'LedgerDB.withTipForker' here, because 'curForker' might be
+    -- returned as part of the selected chain.
+    curForker <- LedgerDB.getForkerAtWellKnownPoint lgrDB rr VolatileTip
 
     chains <- constructChains i succsOf
 
     -- We use the empty fragment anchored at @i@ as the current chain (and
     -- ledger) and the default in case there is no better candidate.
     let curChain          = Empty (AF.castAnchor i)
-        curChainAndLedger = VF.ValidatedFragment curChain ledger
+    curChainAndLedger <- VF.newM curChain curForker
 
     case NE.nonEmpty (filter (preferAnchoredCandidate bcfg curChain) chains) of
       -- If there are no candidates, no chain selection is needed
       Nothing      -> return curChainAndLedger
-      Just chains' -> maybe curChainAndLedger toChainAndLedger <$>
-        chainSelection' curChainAndLedger chains'
+      Just chains' ->
+        chainSelection' curChainAndLedger chains' >>= \case
+          Nothing       -> pure curChainAndLedger
```

Comment:
`chainSelection'` returns a new forker that we want to commit to, so in that we close `curForker` because we won't return it. In case `chainSelection'` returned no new forker, the we return `curForker`. It's [these lines in `openDBInternal`](https://github.com/IntersectMBO/ouroboros-consensus/pull/1267/files#diff-7efe335a37d3acbd318b70c91aa6b26d7e0ff97e485c225771659a15a0ec10b5R157-R176) that will close the forker (be it `curForker` or a new one) eventually.

Arguably, the commit part could be done here in `initialChainSelection` instead of in `openDBInternal`

Reply:
Probably no need. Just clarifying it in a comment is enough.